File: src/simulator.py
"""
Simulation engine for WSN energy routing.

Handles the round-by-round lifecycle:
1. Ambient energy harvesting
2. Cluster head election (LEACH or harvesting-aware)
3. Route calculation (Dijkstra, A*, classical DP, or Time-Augmented DP)
4. Packet transmission, energy deduction, and live rerouting (DSU)
5. Statistics logging and lifetime milestone tracking
"""

import time
import csv
import os
import math
from typing import Dict, List, Tuple, Optional, Union, Any
from network import Node, Graph
from energy_model import EnergyModel
from clustering import leach_clustering, simulate_clustering_round
from routing import dijkstra, astar, compute_routes_for_cluster_heads, compare_dijkstra_astar, rip_up_and_reroute
from dp_lifetime import dp_lifetime_maximin_path, dp_time_augmented_lifetime
from harvesting_model import HarvestingProfile, create_harvesting_model


class Simulator:
    """Discrete-round simulator for wireless sensor networks."""

    def __init__(
        self,
        num_nodes: int = 100,
        area_width: float = 100.0,
        area_height: float = 100.0,
        base_station_pos: Tuple[float, float] = (50.0, 50.0),
        initial_energy: float = 2.0,
        max_battery_capacity: Optional[float] = None,
        desired_clusters_ratio: float = 0.05,
        k_bits: int = 4000,
        data_per_node_per_round: int = 4000,
        aggregation_bits: int = 1000,
        enable_dp_routing: bool = False,
        enable_time_dp: bool = False,
        enable_harvesting_ch: bool = False,
        enable_live_reroute: bool = False,
        harvesting_profile: Optional[Union[HarvestingProfile, str]] = None,
        harvesting_kwargs: Optional[Dict[str, Any]] = None,
        max_dp_hops: int = 5,
        routing_algorithm: str = 'dijkstra',
        transmission_range: Optional[float] = None,
        seed: Optional[int] = None
    ):
        self.num_nodes = num_nodes
        self.area_width = area_width
        self.area_height = area_height
        self.base_station_pos = base_station_pos
        self.initial_energy = initial_energy
        self.max_battery_capacity = max_battery_capacity if max_battery_capacity is not None else max(initial_energy, 2.0)
        self.desired_clusters_ratio = desired_clusters_ratio
        self.k_bits = k_bits
        self.data_per_node_per_round = data_per_node_per_round
        self.aggregation_bits = aggregation_bits
        self.enable_dp_routing = enable_dp_routing
        self.enable_time_dp = enable_time_dp
        self.enable_harvesting_ch = enable_harvesting_ch
        self.enable_live_reroute = enable_live_reroute
        self.max_dp_hops = max_dp_hops
        self.routing_algorithm = routing_algorithm
        self.transmission_range = transmission_range
        self.seed = seed

        # Initialize nodes and network graph
        self.nodes: Dict[int, Node] = {}
        self._create_nodes()
        self.graph = Graph(self.nodes)
        self.energy_model = EnergyModel()

        if isinstance(harvesting_profile, str):
            kwargs = dict(harvesting_kwargs or {})
            if 'nodes' not in kwargs:
                kwargs['nodes'] = self.nodes
            self.harvesting_model: Optional[HarvestingProfile] = create_harvesting_model(harvesting_profile, **kwargs)
        else:
            self.harvesting_model = harvesting_profile

        # Track history for plotting and metrics
        self.round_number = 0
        self.alive_nodes_history: List[int] = []
        self.total_energy_history: List[float] = []
        self.harvested_energy_history: List[float] = []
        self.cluster_heads_history: List[List[int]] = []
        self.routes_history: List[Dict[int, Tuple[Optional[List[int]], float]]] = []
        self.reroute_events_history: List[int] = []
        self.energy_matrix: List[List[float]] = []  # shape: [round, node_index]

        self.first_node_death_round: Optional[int] = None
        self.half_nodes_dead_round: Optional[int] = None
        self.last_node_death_round: Optional[int] = None

        os.makedirs('results', exist_ok=True)

    def _create_nodes(self):
        """Places sensor nodes uniformly in the simulation area."""
        import random
        # Two separate RNG streams seeded deterministically:
        # - placement_rng: controls node (x, y) coordinates
        # - self._rng: controls all per-round stochastic draws (CH election, etc.)
        # Using seed+1 for the second stream keeps them independent.
        placement_rng = random.Random(self.seed) if self.seed is not None else random.Random()
        self._rng: Optional[random.Random] = random.Random(self.seed + 1) if self.seed is not None else None
        for i in range(self.num_nodes):
            x = placement_rng.uniform(0, self.area_width)
            y = placement_rng.uniform(0, self.area_height)
            self.nodes[i] = Node(
                node_id=i,
                x=x,
                y=y,
                initial_energy=self.initial_energy,
                max_energy=self.max_battery_capacity
            )

    def _update_graph_edge_weights(self):
        self.graph.update_edge_weights(self.energy_model)

    def _simulate_round(self) -> bool:
        """Executes one round of harvesting, clustering, routing, and transmission."""
        self.round_number += 1

        # 1. Harvest ambient energy
        round_harvested_total = 0.0
        if self.harvesting_model is not None:
            for nid, node in self.nodes.items():
                if node.is_alive:
                    harvest_amt = self.harvesting_model.sample_harvest(
                        node_id=nid,
                        current_time=self.round_number,
                        duration=1.0
                    )
                    stored = node.harvest_energy(harvest_amt)
                    round_harvested_total += stored
        self.harvested_energy_history.append(round_harvested_total)

        # Record snapshot of current energies for heatmaps
        self.energy_matrix.append([self.nodes[i].residual_energy for i in range(self.num_nodes)])

        # 2. Cluster formation
        ch_model = self.harvesting_model if self.enable_harvesting_ch else None
        cluster_assignment, cluster_heads, _ = simulate_clustering_round(
            nodes=self.nodes,
            energy_model=self.energy_model,
            desired_clusters_ratio=self.desired_clusters_ratio,
            harvesting_model=ch_model,
            current_time=self.round_number,
            lookahead_rounds=1,
            rng=self._rng
        )

        for nid, node in self.nodes.items():
            if not node.is_alive:
                continue
            if nid in cluster_assignment:
                node.cluster_id = cluster_assignment[nid]
                node.role = 'CH' if nid in cluster_heads else 'member'
            else:
                node.role = 'member'
                node.cluster_id = -1

        # 3. Path discovery to base station
        alive_nodes_set = set(self.graph.alive_nodes())
        routes: Dict[int, Tuple[Optional[List[int]], float]] = {}

        if self.enable_time_dp:
            for ch in cluster_heads:
                if ch not in alive_nodes_set:
                    routes[ch] = (None, float('inf'))
                    continue
                lifetime, path, schedule = dp_time_augmented_lifetime(
                    nodes=self.nodes,
                    adj_list=self.graph.adjacency_list,
                    source=ch,
                    base_station_pos=self.base_station_pos,
                    energy_model=self.energy_model,
                    alive_nodes=alive_nodes_set,
                    harvesting_model=self.harvesting_model,
                    current_time=self.round_number,
                    max_hops=self.max_dp_hops,
                    hop_delay=1,
                    k_bits=self.k_bits,
                    transmission_range=self.transmission_range
                )
                routes[ch] = (path, lifetime)

        elif self.enable_dp_routing:
            for ch in cluster_heads:
                if ch not in alive_nodes_set:
                    routes[ch] = (None, float('inf'))
                    continue
                lifetime, path = dp_lifetime_maximin_path(
                    nodes=self.nodes,
                    adj_list=self.graph.adjacency_list,
                    source=ch,
                    base_station_pos=self.base_station_pos,
                    energy_model=self.energy_model,
                    alive_nodes=alive_nodes_set,
                    max_hops=self.max_dp_hops,
                    k_bits=self.k_bits,
                    transmission_range=self.transmission_range
                )
                routes[ch] = (path, lifetime)

        else:
            routes = compute_routes_for_cluster_heads(
                nodes=self.nodes,
                graph=self.graph,
                energy_model=self.energy_model,
                cluster_heads=cluster_heads,
                base_station_pos=self.base_station_pos,
                alive_nodes=alive_nodes_set,
                algorithm=self.routing_algorithm,
                transmission_range=self.transmission_range
            )

        # 4. Data transmission and energy consumption
        members_per_ch: Dict[int, List[int]] = {ch: [] for ch in cluster_heads}
        for nid, ch_id in cluster_assignment.items():
            if nid not in members_per_ch:
                members_per_ch.setdefault(ch_id, []).append(nid)

        # 4A: Member nodes transmit to their respective cluster head
        for nid, node in self.nodes.items():
            if not node.is_alive or node.role != 'member':
                continue
            ch_id = node.cluster_id
            if ch_id != -1 and ch_id in self.nodes and self.nodes[ch_id].is_alive:
                ch_node = self.nodes[ch_id]
                dist = node.distance_to(ch_node)
                e_tx = self.energy_model.transmit_energy(self.data_per_node_per_round, dist)
                node.consume_energy(e_tx)

        # 4B: Cluster heads aggregate and forward data to base station
        round_reroute_count = 0

        for ch_id in cluster_heads:
            ch_node = self.nodes[ch_id]
            if not ch_node.is_alive:
                continue

            num_members = len(members_per_ch.get(ch_id, []))
            if num_members > 0:
                e_rx = self.energy_model.receive_energy(num_members * self.data_per_node_per_round)
                ch_node.consume_energy(e_rx)

            route_info = routes.get(ch_id)
            if not route_info or not route_info[0] or len(route_info[0]) < 2:
                continue

            path = list(route_info[0])
            hop_idx = 0

            while hop_idx < len(path) - 1:
                tx_id = path[hop_idx]
                rx_id = path[hop_idx + 1]

                if tx_id == -1:
                    break

                tx_node = self.nodes[tx_id]
                if not tx_node.is_alive:
                    break

                if rx_id == -1:
                    dist = math.sqrt((tx_node.x - self.base_station_pos[0])**2 + (tx_node.y - self.base_station_pos[1])**2)
                else:
                    rx_node = self.nodes[rx_id]
                    dist = tx_node.distance_to(rx_node)

                    # If next hop node died, attempt local reroute
                    if not rx_node.is_alive or rx_node.residual_energy <= 0.0:
                        if self.enable_live_reroute:
                            new_path, _ = rip_up_and_reroute(
                                nodes=self.nodes,
                                adj_list=self.graph.adjacency_list,
                                failed_node_id=rx_id,
                                active_path=path,
                                base_station_pos=self.base_station_pos,
                                energy_model=self.energy_model,
                                alive_nodes=set(self.graph.alive_nodes()),
                                harvesting_model=self.harvesting_model,
                                current_time=self.round_number,
                                transmission_range=self.transmission_range,
                                k=self.aggregation_bits
                            )
                            if new_path is not None and len(new_path) >= 2:
                                path = new_path
                                round_reroute_count += 1
                                continue
                        break

                # Deduct transmission and reception costs
                e_tx = self.energy_model.transmit_energy(self.aggregation_bits, dist)
                tx_node.consume_energy(e_tx)

                if rx_id != -1:
                    e_rx = self.energy_model.receive_energy(self.aggregation_bits)
                    self.nodes[rx_id].consume_energy(e_rx)

                hop_idx += 1

        self.reroute_events_history.append(round_reroute_count)

        # 5. Record round statistics
        alive_nodes = self.graph.alive_nodes()
        num_alive = len(alive_nodes)
        total_energy = sum(self.nodes[nid].residual_energy for nid in alive_nodes)

        self.alive_nodes_history.append(num_alive)
        self.total_energy_history.append(total_energy)
        self.cluster_heads_history.append(list(cluster_heads))
        self.routes_history.append(routes)

        if self.first_node_death_round is None and num_alive < self.num_nodes:
            self.first_node_death_round = self.round_number
        if self.half_nodes_dead_round is None and num_alive < (self.num_nodes / 2.0):
            self.half_nodes_dead_round = self.round_number
        if self.last_node_death_round is None and num_alive == 0:
            self.last_node_death_round = self.round_number

        return num_alive > 0

    def run(self, max_rounds: int = 1000, verbose: bool = True):
        """Runs the simulation until maximum rounds reached or all nodes die."""
        if verbose:
            print(f"Starting simulation: {self.num_nodes} nodes, max {max_rounds} rounds.")
            print(f"Area: {self.area_width}x{self.area_height}m, Base Station at {self.base_station_pos}")
            harv_desc = str(self.harvesting_model) if self.harvesting_model else "None (No Harvesting)"
            print(f"Harvesting Profile: {harv_desc}")
            mode_desc = "Time-Augmented DP" if self.enable_time_dp else ("Maximin DP" if self.enable_dp_routing else self.routing_algorithm.upper())
            print(f"Routing Mode: {mode_desc} | Harv-Aware CH: {self.enable_harvesting_ch} | Live Reroute: {self.enable_live_reroute}")

        start_time = time.time()
        for _ in range(max_rounds):
            alive = self._simulate_round()
            if not alive:
                if verbose:
                    print(f"Network depleted at round {self.round_number}")
                break
        elapsed = time.time() - start_time

        self._save_simulation_log()

        if verbose:
            print("\n--- Simulation Summary ---")
            print(f"Completed rounds: {self.round_number} in {elapsed:.2f}s")
            print(f"First Node Death (FND): {self.first_node_death_round or 'None'}")
            print(f"Half Nodes Dead (HND):  {self.half_nodes_dead_round or 'None'}")
            print(f"Final Alive Nodes:      {self.alive_nodes_history[-1] if self.alive_nodes_history else 0}/{self.num_nodes}")
            print(f"Final Total Energy:     {self.total_energy_history[-1]:.4f} J" if self.total_energy_history else "0.0 J")

    def _save_simulation_log(self, filepath: str = 'results/simulation_log.csv'):
        """Saves simulation stats to CSV."""
        with open(filepath, 'w', newline='') as csvfile:
            writer = csv.writer(csvfile)
            writer.writerow(['round', 'alive_nodes', 'total_energy_joules', 'harvested_energy_joules', 'reroute_events'])
            for r in range(len(self.alive_nodes_history)):
                alive = self.alive_nodes_history[r]
                energy = self.total_energy_history[r]
                harvest = self.harvested_energy_history[r] if r < len(self.harvested_energy_history) else 0.0
                reroutes = self.reroute_events_history[r] if r < len(self.reroute_events_history) else 0
                writer.writerow([r + 1, alive, energy, harvest, reroutes])


def main():
    import argparse
    parser = argparse.ArgumentParser(description='WSN Energy-Harvesting Routing Simulator')
    parser.add_argument('--nodes', type=int, default=100, help='Number of nodes')
    parser.add_argument('--area', type=float, default=100.0, help='Area width/height')
    parser.add_argument('--bs-x', type=float, default=50.0, help='BS X coordinate')
    parser.add_argument('--bs-y', type=float, default=50.0, help='BS Y coordinate')
    parser.add_argument('--initial-energy', type=float, default=2.0, help='Initial energy per node (J)')
    parser.add_argument('--cluster-ratio', type=float, default=0.05, help='Cluster ratio')
    parser.add_argument('--max-rounds', type=int, default=500, help='Max rounds')
    parser.add_argument('--k-bits', type=int, default=4000, help='Bits per packet')
    parser.add_argument('--data-per-node', type=int, default=4000, help='Data bits per node')
    parser.add_argument('--aggregation-bits', type=int, default=1000, help='Aggregation bits')
    parser.add_argument('--enable-dp', action='store_true', help='Enable classical DP')
    parser.add_argument('--enable-time-dp', action='store_true', help='Enable Time-Augmented DP')
    parser.add_argument('--enable-harvesting-ch', action='store_true', help='Enable harvesting-aware CH rotation')
    parser.add_argument('--enable-live-reroute', action='store_true', help='Enable DSU live reroute')
    parser.add_argument('--harvesting', type=str, default='solar', choices=['solar', 'stochastic', 'constant', 'none', 'shadowed', 'hotspot'], help='Harvesting profile')
    parser.add_argument('--routing', type=str, choices=['dijkstra', 'astar', 'energy_dijkstra'], default='dijkstra', help='Routing algorithm')
    args = parser.parse_args()

    sim = Simulator(
        num_nodes=args.nodes,
        area_width=args.area,
        area_height=args.area,
        base_station_pos=(args.bs_x, args.bs_y),
        initial_energy=args.initial_energy,
        desired_clusters_ratio=args.cluster_ratio,
        k_bits=args.k_bits,
        data_per_node_per_round=args.data_per_node,
        aggregation_bits=args.aggregation_bits,
        enable_dp_routing=args.enable_dp,
        enable_time_dp=args.enable_time_dp,
        enable_harvesting_ch=args.enable_harvesting_ch,
        enable_live_reroute=args.enable_live_reroute,
        harvesting_profile=args.harvesting,
        routing_algorithm=args.routing
    )
    sim.run(max_rounds=args.max_rounds)


if __name__ == "__main__":
    main()

File: frontend/src/engine/simulator.ts
import { Node, NetworkGraph, generateNodeTopology } from './network';
import { EnergyModel } from './energyModel';
import {
  HarvestingProfile,
  ConstantHarvesting,
  SolarHarvesting,
  StochasticPoissonHarvesting,
  RealTraceSolarHarvesting,
  createPRNG
} from './harvestingModel';
import { simulateClusteringRound } from './clustering';
import {
  dijkstra,
  astar,
  dpTimeAugmentedLifetime,
  dpMaximinPath,
  ripUpAndReroute
} from './routing';
import {
  SimulationConfig,
  SimulationResults,
  SimulationSummary,
  TimeSeriesData,
  DetailedData,
  NodePosition
} from '../types';

export class WsnSimulator {
  public config: SimulationConfig;
  public nodes: Map<number, Node>;
  public graph: NetworkGraph;
  public energyModel: EnergyModel;
  public harvestingModel: HarvestingProfile | null;
  private rng: () => number;

  public roundNumber: number = 0;
  public aliveNodesHistory: number[] = [];
  public totalEnergyHistory: number[] = [];
  public harvestedEnergyHistory: number[] = [];
  public consumedEnergyHistory: number[] = [];
  public rerouteEventsHistory: number[] = [];
  public fairnessIndexHistory: number[] = [];
  public pdrHistory: number[] = [];
  public clusterHeadsHistory: number[][] = [];
  public clusterAssignmentsHistory: Record<number, number>[] = [];
  public routesHistory: Record<number, [number[] | null, number]>[] = [];
  public energyMatrix: number[][] = []; // [round][nodeId]

  public firstNodeDeathRound: number | null = null;
  public halfNodesDeadRound: number | null = null;
  public lastNodeDeathRound: number | null = null;

  constructor(config: SimulationConfig) {
    this.config = { ...config };
    this.rng = createPRNG(this.config.seed + 1);
    this.energyModel = new EnergyModel();

    // 1. Generate node topology
    this.nodes = generateNodeTopology(
      this.config.nodes,
      this.config.area,
      this.config.area,
      this.config.init_energy,
      this.config.max_capacity,
      this.config.topology_distribution || 'uniform',
      this.config.harvesting_profile === 'shadowed_solar' ? (this.config.shadow_fraction || 0.35) : 0.0,
      this.config.seed
    );

    this.graph = new NetworkGraph(this.nodes);
    this.graph.buildGraph(this.config.transmission_range, this.energyModel);

    // 2. Initialize harvesting model
    this.harvestingModel = this.createHarvestingModel();
  }

  private createHarvestingModel(): HarvestingProfile | null {
    const profile = this.config.harvesting_profile;
    const seed = this.config.seed;

    if (profile === 'none') {
      return null;
    } else if (profile === 'constant') {
      return new ConstantHarvesting(this.config.constant_rate || 0.005);
    } else if (profile === 'solar') {
      return new SolarHarvesting(
        this.config.solar_peak || 0.03,
        this.config.solar_period || 24,
        this.config.solar_day_fraction || 0.5,
        seed
      );
    } else if (profile === 'shadowed_solar') {
      const solar = new SolarHarvesting(
        this.config.solar_peak || 0.03,
        this.config.solar_period || 24,
        this.config.solar_day_fraction || 0.5,
        seed
      );
      // Map spatial shadow multiplier onto nodes
      for (const [id, node] of this.nodes.entries()) {
        solar.setNodeMultiplier(id, node.shadowMultiplier);
      }
      return solar;
    } else if (profile === 'stochastic') {
      return new StochasticPoissonHarvesting(
        this.config.stoch_lambda || 2.0,
        this.config.stoch_quantum || 0.005,
        seed
      );
    } else if (profile === 'trace') {
      return new RealTraceSolarHarvesting(this.config.solar_peak || 0.03, seed);
    }
    return null;
  }

  /**
   * Jain's Fairness Index: J = (sum x_i)^2 / (n * sum x_i^2)
   */
  private calculateFairnessIndex(): number {
    const energies = Array.from(this.nodes.values()).map(n => n.residualEnergy);
    const n = energies.length;
    if (n === 0) return 1.0;
    const sum = energies.reduce((a, b) => a + b, 0);
    const sumSq = energies.reduce((a, b) => a + b * b, 0);
    if (sumSq <= 1e-9) return 1.0;
    return (sum * sum) / (n * sumSq);
  }

  public simulateRound(): boolean {
    this.roundNumber++;
    const kBits = this.config.k_bits || 4000;
    const dataPerNode = this.config.data_per_node_per_round || 4000;
    const aggBits = this.config.aggregation_bits || 1000;
    const baseStationPos: [number, number] = [this.config.bs_x, this.config.bs_y];

    // 1. Ambient Energy Harvesting Step
    let roundHarvestedTotal = 0;
    if (this.harvestingModel) {
      for (const node of this.nodes.values()) {
        if (node.isAlive) {
          const harvestAmt = this.harvestingModel.sampleHarvest(node.id, this.roundNumber, 1.0);
          const stored = node.harvestEnergy(harvestAmt);
          roundHarvestedTotal += stored;
        }
      }
    }
    this.harvestedEnergyHistory.push(roundHarvestedTotal);

    // Record energy matrix snapshot
    const energySnapshot: number[] = [];
    for (let i = 0; i < this.config.nodes; i++) {
      const node = this.nodes.get(i);
      energySnapshot.push(node ? node.residualEnergy : 0.0);
    }
    this.energyMatrix.push(energySnapshot);

    // 2. Clustering Step (LEACH or Harvesting-Aware EH-LEACH)
    const chHarvestModel = !this.config.disable_harvesting_ch ? this.harvestingModel : null;
    const { clusterAssignment, clusterHeads, membersPerCH } = simulateClusteringRound(
      this.nodes,
      this.energyModel,
      this.config.cluster_ratio,
      chHarvestModel,
      this.roundNumber,
      1,
      this.rng
    );

    this.clusterHeadsHistory.push([...clusterHeads]);
    this.clusterAssignmentsHistory.push({ ...clusterAssignment });

    // 3. Path Discovery to Base Station
    const aliveNodesSet = new Set<number>();
    for (const [id, node] of this.nodes.entries()) {
      if (node.isAlive) aliveNodesSet.add(id);
    }

    const routes: Record<number, [number[] | null, number]> = {};

    for (const ch of clusterHeads) {
      if (!aliveNodesSet.has(ch)) {
        routes[ch] = [null, Infinity];
        continue;
      }

      if (!this.config.disable_time_dp) {
        // Time-Augmented DP
        const { bottleneckEnergy, path } = dpTimeAugmentedLifetime(
          this.nodes,
          ch,
          baseStationPos,
          this.energyModel,
          aliveNodesSet,
          this.harvestingModel,
          this.roundNumber,
          this.config.max_dp_hops || 5,
          1,
          kBits,
          this.config.transmission_range
        );
        routes[ch] = [path, bottleneckEnergy];
      } else if (this.config.routing_algorithm === 'dp_maximin') {
        // Maximin DP
        const { bottleneckEnergy, path } = dpMaximinPath(
          this.nodes,
          ch,
          baseStationPos,
          this.energyModel,
          aliveNodesSet,
          this.config.max_dp_hops || 5,
          kBits,
          this.config.transmission_range
        );
        routes[ch] = [path, bottleneckEnergy];
      } else if (this.config.routing_algorithm === 'astar') {
        // A* Pathfinding
        const { cost, path } = astar(
          this.nodes,
          this.graph,
          ch,
          baseStationPos,
          this.energyModel,
          aliveNodesSet,
          kBits,
          this.config.transmission_range
        );
        routes[ch] = [path, cost];
      } else if (this.config.routing_algorithm === 'energy_dijkstra') {
        // Energy-Aware Dijkstra
        const { cost, path } = dijkstra(
          this.nodes,
          this.graph,
          ch,
          baseStationPos,
          this.energyModel,
          aliveNodesSet,
          true,
          kBits,
          this.config.transmission_range
        );
        routes[ch] = [path, cost];
      } else {
        // Standard Dijkstra (Min Energy)
        const { cost, path } = dijkstra(
          this.nodes,
          this.graph,
          ch,
          baseStationPos,
          this.energyModel,
          aliveNodesSet,
          false,
          kBits,
          this.config.transmission_range
        );
        routes[ch] = [path, cost];
      }
    }

    this.routesHistory.push(routes);

    // 4. Data Transmission & Energy Deductions
    let roundConsumedTotal = 0;
    let packetsDelivered = 0;
    let packetsGenerated = 0;

    // 4A: Member nodes transmit to their elected Cluster Head
    for (const [nid, node] of this.nodes.entries()) {
      if (!node.isAlive || node.role !== 'member') continue;
      const chId = node.clusterId;
      if (chId !== -1 && this.nodes.has(chId) && this.nodes.get(chId)!.isAlive) {
        const chNode = this.nodes.get(chId)!;
        const dist = node.distanceTo(chNode);
        const eTx = this.energyModel.transmitEnergy(dataPerNode, dist);
        roundConsumedTotal += node.consumeEnergy(eTx);
        packetsGenerated++;
      }
    }

    // 4B: Cluster Heads aggregate and forward packets along multi-hop routes
    let roundRerouteCount = 0;

    for (const chId of clusterHeads) {
      const chNode = this.nodes.get(chId);
      if (!chNode || !chNode.isAlive) continue;

      const memberCount = (membersPerCH[chId] || []).length;
      if (memberCount > 0) {
        // Aggregation and reception cost
        const eRx = this.energyModel.receiveEnergy(memberCount * dataPerNode);
        const eAgg = this.energyModel.aggregationEnergy(aggBits, memberCount);
        roundConsumedTotal += chNode.consumeEnergy(eRx + eAgg);
      }

      const routeInfo = routes[chId];
      if (!routeInfo || !routeInfo[0] || routeInfo[0].length < 2) {
        continue;
      }

      let path = [...routeInfo[0]];
      let hopIdx = 0;
      let pathSucceeded = false;

      while (hopIdx < path.length - 1) {
        const txId = path[hopIdx];
        const rxId = path[hopIdx + 1];

        if (txId === -1) break;
        const txNode = this.nodes.get(txId);
        if (!txNode || !txNode.isAlive) break;

        let dist = 0;
        if (rxId === -1) {
          // Transmission to Base Station
          dist = Math.sqrt((txNode.x - baseStationPos[0]) ** 2 + (txNode.y - baseStationPos[1]) ** 2);
        } else {
          const rxNode = this.nodes.get(rxId);
          if (!rxNode || !rxNode.isAlive || rxNode.residualEnergy <= 1e-9) {
            // Next-hop relay is dead mid-transmission! Attempt DSU live detour recovery
            if (!this.config.disable_live_reroute) {
              const currentAliveSet = new Set(this.graph.getAliveNodeIds());
              const detour = ripUpAndReroute(
                this.nodes,
                rxId,
                path,
                baseStationPos,
                this.energyModel,
                currentAliveSet,
                aggBits,
                this.config.transmission_range
              );

              if (detour.success && detour.newPath && detour.newPath.length >= 2) {
                path = detour.newPath;
                roundRerouteCount++;
                continue; // Retry transmission from current hop with updated detour path
              }
            }
            break; // Detour failed, packet dropped
          }
          dist = txNode.distanceTo(rxNode);
        }

        // Deduct transmission energy from transmitter
        const eTx = this.energyModel.transmitEnergy(aggBits, dist);
        roundConsumedTotal += txNode.consumeEnergy(eTx);

        // Deduct reception energy from receiver if not Base Station
        if (rxId !== -1) {
          const rxNode = this.nodes.get(rxId)!;
          const eRx = this.energyModel.receiveEnergy(aggBits);
          roundConsumedTotal += rxNode.consumeEnergy(eRx);
        } else {
          pathSucceeded = true;
        }

        hopIdx++;
      }

      if (pathSucceeded) {
        packetsDelivered += Math.max(1, memberCount);
      }
    }

    this.consumedEnergyHistory.push(roundConsumedTotal);
    this.rerouteEventsHistory.push(roundRerouteCount);

    // 5. Round Statistics and Milestone Tracking
    const aliveIds = this.graph.getAliveNodeIds();
    const numAlive = aliveIds.length;
    let totalEnergy = 0;
    for (const nid of aliveIds) {
      totalEnergy += this.nodes.get(nid)!.residualEnergy;
    }

    this.aliveNodesHistory.push(numAlive);
    this.totalEnergyHistory.push(totalEnergy);
    this.fairnessIndexHistory.push(this.calculateFairnessIndex());

    const pdr = packetsGenerated > 0 ? Math.min(1.0, packetsDelivered / packetsGenerated) : (numAlive > 0 ? 1.0 : 0.0);
    this.pdrHistory.push(pdr);

    if (this.firstNodeDeathRound === null && numAlive < this.config.nodes) {
      this.firstNodeDeathRound = this.roundNumber;
    }
    if (this.halfNodesDeadRound === null && numAlive <= Math.floor(this.config.nodes / 2)) {
      this.halfNodesDeadRound = this.roundNumber;
    }
    if (this.lastNodeDeathRound === null && numAlive === 0) {
      this.lastNodeDeathRound = this.roundNumber;
    }

    return numAlive > 0;
  }

  public run(): SimulationResults {
    const startTime = performance.now();
    const maxRounds = this.config.rounds || 200;

    for (let r = 0; r < maxRounds; r++) {
      const stillAlive = this.simulateRound();
      if (!stillAlive) break;
    }

    const executionTimeMs = performance.now() - startTime;

    // Compile summary
    const completedRounds = this.roundNumber;
    const finalAlive = this.aliveNodesHistory[this.aliveNodesHistory.length - 1] || 0;
    const finalEnergy = this.totalEnergyHistory[this.totalEnergyHistory.length - 1] || 0;
    const totalHarvested = this.harvestedEnergyHistory.reduce((a, b) => a + b, 0);
    const totalConsumed = this.consumedEnergyHistory.reduce((a, b) => a + b, 0);
    const totalReroutes = this.rerouteEventsHistory.reduce((a, b) => a + b, 0);
    const avgPdr = this.pdrHistory.length > 0 ? this.pdrHistory.reduce((a, b) => a + b, 0) / this.pdrHistory.length : 0;
    const finalFairness = this.fairnessIndexHistory[this.fairnessIndexHistory.length - 1] || 1.0;

    const summary: SimulationSummary = {
      completed_rounds: completedRounds,
      first_node_death_round: this.firstNodeDeathRound,
      half_nodes_dead_round: this.halfNodesDeadRound,
      last_node_death_round: this.lastNodeDeathRound,
      final_alive_nodes: finalAlive,
      total_nodes: this.config.nodes,
      final_total_energy: Number(finalEnergy.toFixed(5)),
      total_harvested_energy: Number(totalHarvested.toFixed(5)),
      total_consumed_energy: Number(totalConsumed.toFixed(5)),
      total_reroutes: totalReroutes,
      network_lifetime_efficiency: Number(((finalAlive / this.config.nodes) * 100).toFixed(1)),
      average_pdr: Number((avgPdr * 100).toFixed(1)),
      jains_fairness_final: Number(finalFairness.toFixed(3)),
      execution_time_ms: Math.round(executionTimeMs)
    };

    const timeSeries: TimeSeriesData = {
      rounds: Array.from({ length: completedRounds }, (_, i) => i + 1),
      alive_nodes: this.aliveNodesHistory,
      total_energy: this.totalEnergyHistory,
      harvested_energy: this.harvestedEnergyHistory,
      consumed_energy: this.consumedEnergyHistory,
      reroute_events: this.rerouteEventsHistory,
      fairness_index: this.fairnessIndexHistory,
      pdr_history: this.pdrHistory
    };

    const nodePositions: Record<string, NodePosition> = {};
    const nodeShadowMultipliers: Record<string, number> = {};
    for (const [id, node] of this.nodes.entries()) {
      nodePositions[String(id)] = { x: Number(node.x.toFixed(2)), y: Number(node.y.toFixed(2)) };
      nodeShadowMultipliers[String(id)] = node.shadowMultiplier;
    }

    const detailedData: DetailedData = {
      energy_matrix: this.energyMatrix,
      cluster_heads_history: this.clusterHeadsHistory,
      cluster_assignments_history: this.clusterAssignmentsHistory,
      routes_history: this.routesHistory,
      node_positions: nodePositions,
      node_shadow_multipliers: nodeShadowMultipliers,
      base_station_position: [this.config.bs_x, this.config.bs_y],
      fnd_round: this.firstNodeDeathRound,
      hnd_round: this.halfNodesDeadRound
    };

    return {
      summary,
      time_series: timeSeries,
      detailed_data: detailedData,
      configuration: this.config
    };
  }
}
